src/lib.rs

```diff
@@ -675,13 +930,191 @@ impl Contract {
+            let review_status = proposal_timeline.get_review_status().clone();
+            self.edit_proposal_timeline(proposal_id, TimelineStatus::Cancelled(review_status));
+        }
+
+        for proposal_id in proposals_to_unlink {
+            self.edit_proposal_linked_rfp(proposal_id, None);
+        }
+
+        self.edit_rfp_timeline(id, RFPTimelineStatus::Cancelled)
+    }
+
+    #[payable]
+    pub fn edit_rfp_timeline(&mut self, id: RFPId, timeline: RFPTimelineStatus) -> Promise {
+        near_sdk::log!("edit_rfp_timeline");
+        let rfp: RFP =
+            self.rfps.get(id.into()).unwrap_or_else(|| panic!("RFP id {} not found", id)).into();
+        let mut body = rfp.snapshot.body.latest_version();
+        body.timeline = timeline;
+
+        self.edit_rfp_internal(id, body.into(), rfp.snapshot.labels)
+    }
+
+    fn edit_rfp_internal(
+        &mut self,
+        id: RFPId,
+        body: VersionedRFPBody,
+        labels: HashSet<String>,
+    ) -> Promise {
+        let editor_id: AccountId = env::predecessor_account_id();
+        require!(
+            self.is_allowed_to_write_rfps(editor_id.clone()),
+            "The account is not allowed to edit RFPs"
+        );
+
+        let mut rfp: RFP =
+            self.rfps.get(id.into()).unwrap_or_else(|| panic!("RFP id {} not found", id)).into();
+
+        let rfp_body = body.clone().latest_version();
+
+        if rfp_body.timeline.is_proposal_selected() {
+            let mut has_approved_proposal = false;
+            for proposal_id in self.get_rfp_linked_proposals(id) {
+                let proposal: Proposal = self
+                    .proposals
+                    .get(proposal_id.into())
+                    .unwrap_or_else(|| panic!("Proposal id {} not found", proposal_id))
+                    .into();
+                if proposal.snapshot.body.latest_version().timeline.is_approved() {
+                    has_approved_proposal = true;
+                    break;
+                }
+            }
+            require!(has_approved_proposal, "Cannot change RFP status to Proposal Selected without an approved proposal linked to this RFP");
+        }
+
+        let old_snapshot = rfp.snapshot.clone();
+        let old_labels_set = old_snapshot.labels.clone();
+        let new_labels = labels;
+        let new_snapshot = RFPSnapshot {
+            editor_id: env::predecessor_account_id(),
+            timestamp: env::block_timestamp(),
+            labels: new_labels.clone(),
+            body: body,
+            linked_proposals: old_snapshot.linked_proposals.clone(),
+        };
+        rfp.snapshot = new_snapshot;
+        rfp.snapshot_history.push(old_snapshot);
+        self.rfps.replace(id.try_into().unwrap(), &rfp.clone().into());
+
+        // Update labels index.
+        let new_labels_set = new_labels;
+
+        let mut edit_proposal_promise: Option<Promise> = None;
+
+        if !old_labels_set.eq(&new_labels_set.clone()) {
```

Comment:
I think you should be able to just compare using `!=` operator since it uses `eq` method under the hood:

```suggestion
        if old_labels_set != new_labels_set {
```